bin/propolis-server/src/lib/migrate/compat.rs

```diff
@@ -0,0 +1,791 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Checks for compatibility of two instance specs.
+
+use std::collections::HashMap;
+
+use crate::spec::{self, SerialPortDevice};
+
+use propolis_api_types::instance_spec::{
+    components::{
+        board::Chipset,
+        devices::{PciPciBridge, SerialPortNumber},
+    },
+    PciPath,
+};
+use thiserror::Error;
+
+trait CompatComponent {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), ComponentIncompatibility>;
+}
+
+trait CompatCollection {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), CollectionIncompatibility>;
+}
```

Comment:
not particularly important, but it occurs to me that this _could_ be one trait, like:
```suggestion
trait Compat {
    type Error;
    fn is_compatible_with(
        &self,
        other: &Self,
    ) -> Result<(), Self::Error>;
}
```

since the methods are otherwise the same. but, the current thing is also fine.